sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpConsumer.cs

```diff
@@ -236,52 +236,56 @@ internal class AmqpConsumer : TransportConsumer
                         link = await ReceiveLink.GetOrCreateAsync(UseMinimum(ConnectionScope.SessionTimeout, tryTimeout)).ConfigureAwait(false);
                         cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
 
-                        var messagesReceived = await Task.Factory.FromAsync
+                        var messagesReceived = await Task.Factory.FromAsync<(ReceivingAmqpLink, int, TimeSpan), IEnumerable<AmqpMessage>>
                         (
-                            (callback, state) => link.BeginReceiveMessages(maximumEventCount, waitTime, callback, state),
-                            (asyncResult) => link.EndReceiveMessages(asyncResult, out amqpMessages),
-                            TaskCreationOptions.RunContinuationsAsynchronously
+                            static (arguments, callback, state) =>
+                            {
+                                var (link, maximumEventCount, waitTime) = arguments;
+                                return link.BeginReceiveMessages(maximumEventCount, waitTime, callback, link);
+                            },
+                            static asyncResult =>
+                            {
+                                var link = (ReceivingAmqpLink)asyncResult.AsyncState;
+                                var received = link.EndReceiveMessages(asyncResult, out var amqpMessages);
+
+                                return received ? amqpMessages : Enumerable.Empty<AmqpMessage>();
+                            },
+                            (link, maximumEventCount, waitTime),
+                            default
                         ).ConfigureAwait(false);
 
                         cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
 
                         // If event messages were received, then package them for consumption and
                         // return them.
 
-                        if ((messagesReceived) && (amqpMessages != null))
+                        foreach (AmqpMessage message in messagesReceived)
```

Comment:
I was just curious about the perf impacts of removing the conditional check before doing the `foreach` and it's quite a bit slower (though very fast in absolute time)

```
|                 Method |     Mean |     Error |    StdDev | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |---------:|----------:|----------:|------:|--------:|------:|------:|------:|----------:|
| ConditionalEnumeration | 1.132 ns | 0.0147 ns | 0.0137 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|          DoEnumeration | 5.490 ns | 0.0459 ns | 0.0429 ns |  4.85 |    0.08 |     - |     - |     - |         - |
```

Reply:
Interesting.

Reply:
I tweaked to return `null` when nothing was received and added a check to bail out early if we have none.